Update QUIC man7 pages for server functionality

andrewkdinh opened the pull request:
Primarily added functions related to QUIC server functionality to the man7 pages about quic.

Fixes [openssl/projects#871](https://github.com/openssl/project/issues/871)

##### Checklist

- [x] documentation is added or updated

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -7,15 +7,17 @@ openssl-quic - OpenSSL QUIC
 OpenSSL 3.2 and later features support for the QUIC transport protocol.
-Currently, only client connectivity is supported. This man page describes the
-usage of QUIC client functionality for both existing and new applications.
+Initially, only client connectivity was supported. Now, server connectivity is
+also supported. This man page describes the usage of QUIC functionality for both
+existing and new applications.
```
I would modify this piece of text. It will very quickly become out of date and irrelevant. In a couple of years time, who cares that we launched with only client support? I don't think we need to mention it.

```suggestion
You can use OpenSSL's QUIC capabilities for both client and server applications.
This man page describes the usage of QUIC functionality for both existing and
new applications.
```

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -324,12 +326,16 @@ determine what changes they will need to make:
 
-An application wishing to use QUIC must use L<OSSL_QUIC_client_method(3)> or
-L<OSSL_QUIC_client_thread_method(3)> as its SSL method. For more information
+An client application wishing to use QUIC must use L<OSSL_QUIC_client_method(3)>
```
```suggestion
A client application wishing to use QUIC must use L<OSSL_QUIC_client_method(3)>
```

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -498,6 +504,57 @@ The following SSL APIs are new but relevant to both QUIC and DTLS:
 
 =over 4
 
+=item L<SSL_new_listener(3)>
```
This function and the others below are in the wrong list. This list is preceded by the text "The following SSL APIs are new but relevant to both QUIC and DTLS". But all of these currently only work for QUIC not DTLS (feasibly we could extend them for DTLS in the future). There is a later list with the text "The following SSL APIs are specific to QUIC". This documentation should be moved there instead.

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -498,6 +504,57 @@ The following SSL APIs are new but relevant to both QUIC and DTLS:
+=item L<SSL_new_listener_from(3)>
+
+Creates a listener SSL object which is subordinate to a QUIC domain SSL object
+I<ssl>. See L<SSL_new_domain(3)> and L<openssl-quic-concurrency(7)> for details
```
Shouldn't `SSL_new_domain` appear in this list somewhere too?

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -498,6 +504,57 @@ The following SSL APIs are new but relevant to both QUIC and DTLS:
+
+=item L<SSL_is_listener(3)>
+
+Returns 1 if and only if a SSL object is a listener SSL object.
```
```suggestion
Returns 1 if and only if an SSL object is a listener SSL object.
```

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -498,6 +504,57 @@ The following SSL APIs are new but relevant to both QUIC and DTLS:
+Returns an informational value listing the number of connections waiting to be
+popped from the queue via calls to SSL_accept_connection().
+
+=item L<SSL_new_from_listener(3)>
```
Note that this function appears to be currently unimplemented (it always returns NULL). I hope there is an issue somewhere to implement it...!!

andrewkdinh replied:
I couldnt find an issue tracking this, so I've opened one to track it: https://github.com/openssl/openssl/issues/25851

Sashan reviewed (changes requested):
I've just pushed some changes to this PR. I'm sorry if it was appropriate I'm still learning git/github. My biggest concern I feel we should address here is 'default' stream mode vs. 'multi-stream' mode. The current naming is unfortunate and leads to confusion. I'm proposing to use 'implicit stream mode' for what we currently call 'default mode' and 'explicit stream mode' for what we currently call 'multi-stream' mode.

The other changes I'm proposing are just nits. for example wording around poll descirpors can be simplified.

But we need to reach wider agreement on my proposals.

Sashan commented:
> I've just pushed some changes to this PR. I'm sorry if it was appropriate I'm still learning git/github. My biggest concern I feel we should address here is 'default' stream mode vs. 'multi-stream' mode. The current naming is unfortunate and leads to confusion. I'm proposing to use 'implicit stream mode' for what we currently call 'default mode' and 'explicit stream mode' for what we currently call 'multi-stream' mode.
> 
> The other changes I'm proposing are just nits. for example wording around poll descirpors can be simplified.
> 
> But we need to reach wider agreement on my proposals.

@andrewkdinh please blackout my commit. I was kind of hoping it will require your intervention to merge those changes to your PR. I'm still github newbie. sorry about that.

Earlier today at stand up @t8m and @mattcaswell briefly discussed my suggestion to rename `default` connection mode to `implicit` connection mode and `multistream` connection mode to explicit connection mode. Unfortunately the current naming (which I don't like much) goes much deeper in code base and it's too late to change it (or more work thought is needed on how to do change it).

andrewkdinh commented:
@Sashan I've reverted your commit. I agree that we should rename default to implicit, but we can save that for another time.

For both @Sashan and @mattcaswell: I've addressed all your PR comments, making sure to clarify the differences between QUIC client and server implementations.

andrewkdinh commented:
@Sashan I've reverted your commit. I agree that we should rename default to implicit, but we can save that for another time.

For both @Sashan and @mattcaswell: I've addressed all your PR comments, making sure to clarify the differences between QUIC client and server implementations.

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -7,42 +7,30 @@ openssl-quic - OpenSSL QUIC
-Currently, only client connectivity is supported. This man page describes the
-usage of QUIC client functionality for both existing and new applications.
+You can use OpenSSL's QUIC capabilities for both client and server applications.
+This man page describes how to let applications use QUIC protocol using the
```
the QUIC protocol

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -7,42 +7,30 @@ openssl-quic - OpenSSL QUIC
+libssl API.
 
-QUIC functionality uses the standard SSL API. A QUIC connection is represented
+QUIC protocol maps to the standard SSL API. A QUIC connection is represented
```
The QUIC protocol

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -7,42 +7,30 @@ openssl-quic - OpenSSL QUIC
-applications. New applications using QUIC are recommended to disable default
-stream mode and use the multi-stream API; see the MODES OF OPERATION section and
-the RECOMMENDATIONS FOR NEW APPLICATIONS section for more information.
+are needed to existing applications which use libssl API to bring QUIC protocol
```
QUIC protocol support in

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -526,6 +533,72 @@ The following SSL APIs are specific to QUIC:
+
+=item L<SSL_is_domain(3)>
+
+Returns 1 if a SSL object is a QUIC domain SSL object.
```
an SSL object

mattcaswell commented on doc/man7/openssl-quic.pod:
```diff
@@ -526,6 +533,72 @@ The following SSL APIs are specific to QUIC:
+
+=item L<SSL_get0_domain(3)>
+
+SSL_get0_domain() obtains a pointer to the QUIC domain SSL object in a SSL
```
an SSL object hierarchy

mattcaswell reviewed (changes requested):
Looks good. A few minor nits remain

andrewkdinh commented:
@mattcaswell For the grammar fixing of "a SSL" to "an SSL", there's a dozen or so occurrences throughout the codebase. We probably don't want to do it in this PR, but I can open a quick PR to fix the rest.

Sashan reviewed (approved):
I have no further comments/suggestions.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Squashed and merged to the feature branch. Thank you.
